File: shared/geocoding.py
import time
import requests
from shared.config import get_osm_user_agent


def reverse_geocode(lat: float, lng: float) -> tuple:
    """Best-effort city+country lookup via Nominatim. Returns (city, country) tuple."""
    try:
        resp = requests.get(
            "https://nominatim.openstreetmap.org/reverse",
            params={"lat": lat, "lon": lng, "format": "json"},
            headers={"User-Agent": get_osm_user_agent()},
            timeout=10,
        )
        if resp.status_code == 200:
            address = resp.json().get("address", {})
            city = (
                address.get("city") or address.get("town") or
                address.get("village") or address.get("hamlet") or ""
            )
            country = address.get("country", "")
            return city, country
    except requests.RequestException as e:
        print(f"Warning: reverse geocode failed for ({lat}, {lng}): {e}")
    return "", ""

File: android/src/services/geocoding.ts
let _lastGeocodeTime = 0;

export async function reverseGeocode(
  lat: number,
  lng: number,
  osmUserAgent: string,
): Promise<[string, string]> {
  const now = Date.now();
  const wait = 1100 - (now - _lastGeocodeTime);
  if (wait > 0) await new Promise<void>(r => setTimeout(r, wait));
  _lastGeocodeTime = Date.now();

  try {
    const res = await fetch(
      `https://nominatim.openstreetmap.org/reverse?lat=${lat}&lon=${lng}&format=json`,
      {
        headers: {
          'User-Agent': osmUserAgent || 'strava2earth/1.0',
          'Accept-Language': 'en',
        },
      },
    );
    if (!res.ok) return ['', ''];
    const data = await res.json();
    const addr = data.address ?? {};
    const city =
      addr.city ?? addr.town ?? addr.village ?? addr.county ?? addr.municipality ?? '';
    const country = addr.country ?? '';
    return [city, country];
  } catch {
    return ['', ''];
  }
}
